core/src/test/scala/org/apache/spark/DataPropertyAccumulatorSuite.scala

```diff
@@ -0,0 +1,383 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.ref.WeakReference
+
+import org.scalatest.Matchers
+
+import org.apache.spark.scheduler._
+import org.apache.spark.util.{AccumulatorContext, AccumulatorMetadata, AccumulatorV2, LongAccumulator}
+
+
+class DataPropertyAccumulatorSuite extends SparkFunSuite with Matchers with LocalSparkContext {
+  test("two partition old and new") {
+    sc = new SparkContext("local[2]", "test")
+    val acc = sc.longAccumulator(dataProperty = true, "l2")
+
+    val a = sc.parallelize(1 to 20, 2)
+    val b = a.map{x => acc.add(x); x}
+    b.cache()
+    b.count()
+    acc.value should be (210)
+  }
+
+  test("single partition") {
+    sc = new SparkContext("local[2]", "test")
+    val acc = sc.longAccumulator(dataProperty = true)
+
+    val a = sc.parallelize(1 to 20, 1)
+    val b = a.map{x => acc.add(x); x}
+    b.cache()
+    b.count()
+    acc.value should be (210)
+  }
+
+  test("adding only the first element per partition should work even if partition is empty") {
+    sc = new SparkContext("local[2]", "test")
+    val acc = sc.longAccumulator(dataProperty = true)
+    val a = sc.parallelize(1 to 20, 30)
+    val b = a.mapPartitions{itr =>
+      acc.add(1)
+      itr
+    }
+    b.count()
+    acc.value should be (30)
+  }
+
+  test("shuffled (combineByKey)") {
+    sc = new SparkContext("local[2]", "test")
+    val a = sc.parallelize(1L to 40L, 5)
+    val buckets = 4
+    val b = a.map{x => ((x % buckets), x)}
+    val inputs = List(b, b.repartition(10), b.partitionBy(new HashPartitioner(5))).map(_.cache())
+    val mapSideCombines = List(true, false)
+    inputs.foreach { input =>
+      mapSideCombines.foreach { mapSideCombine =>
+        val accs = (1 to 4).map(x => sc.longAccumulator(dataProperty = true)).toList
+        val raccs = (1 to 4).map(x => sc.longAccumulator(dataProperty = false)).toList
+        val List(acc, acc1, acc2, acc3) = accs
+        val List(racc, racc1, racc2, racc3) = raccs
+        val c = input.combineByKey(
+          (x: Long) => {acc1.add(1); acc.add(1); racc1.add(1); racc.add(1); x},
+          {(a: Long, b: Long) => acc2.add(1); acc.add(1); racc2.add(1); racc.add(1); (a + b)},
+          {(a: Long, b: Long) => acc3.add(1); acc.add(1); racc3.add(1); racc.add(1); (a + b)},
+          new HashPartitioner(10),
+          mapSideCombine)
+        val d = input.combineByKey(
+          (x: Long) => {acc1.add(1); acc.add(1); x},
+          {(a: Long, b: Long) => acc2.add(1); acc.add(1); (a + b)},
+          {(a: Long, b: Long) => acc3.add(1); acc.add(1); (a + b)},
+          new HashPartitioner(2),
+          mapSideCombine)
+        val e = d.map{x => acc.add(1); x}
```

Comment:
Well there are two ways it could fail, `e.count()` look like duplicates from d - or if the `insertAll` wasn't happening the updates from the read side of `d` would look like they were coming from `e` and we would get double counting.